og_image_writer/src/img.rs

```diff
@@ -54,153 +53,130 @@ pub(super) fn open_and_resize_with_data(
     Ok(ImageInfo(buffer, Size { height, width }))
 }
 
-// see: https://stackoverflow.com/questions/48478497/javascript-gecko-border-radius-adaptation-on-html-canvas-css-border-radius
-// fn calculate_border_radius(r: &mut BorderRadius, w:f32, h: f32) {
-//     let BorderRadius(tl, tr, bl, br) = r;
-//     let max_radius_width = cmp::max(*tl + *tr, *bl + *br) as f32;
-//     let max_radius_height = cmp::max(*tl + *bl, *tr + *br) as f32;
-//     let width_ratio = w / max_radius_width;
-//     let height_ratio = h / max_radius_height;
-//     let scale_ratio = f32::min(f32::min(width_ratio, height_ratio), 1.);
-
-//     *tl = (*tl as f32 * scale_ratio) as u32;
-//     *tr = (*tr as f32 * scale_ratio) as u32;
-//     *bl = (*tr as f32 * scale_ratio) as u32;
-//     *br = (*tr as f32 * scale_ratio) as u32;
-// }
-
-fn while_radius<F>(img: &mut ImageBuffer<Rgba<u8>, Vec<u8>>, r: i32, size: (i32, i32), draw: F)
-where
-    F: Fn(&mut ImageBuffer<Rgba<u8>, Vec<u8>>, (i32, i32), (i32, i32), Rgba<u8>),
-{
-    if r <= 0 {
+// See https://users.rust-lang.org/t/how-to-trim-image-to-circle-image-without-jaggy/70374
+// Thanks @steffahn for sending some ideas.
+pub(super) fn round(img: &mut ImageBuffer<Rgba<u8>, Vec<u8>>, radius: &mut BorderRadius) {
+    let (width, height) = img.dimensions();
+    assert!(radius.0 + radius.1 <= width);
+    assert!(radius.3 + radius.2 <= width);
+    assert!(radius.0 + radius.3 <= height);
+    assert!(radius.1 + radius.2 <= height);
+
+    // top left
+    border_radius(img, radius.0, |x, y| (x - 1, y - 1));
+    // top right
+    border_radius(img, radius.1, |x, y| (width - x, y - 1));
+    // bottom right
+    border_radius(img, radius.2, |x, y| (width - x, height - y));
+    // bottom left
+    border_radius(img, radius.3, |x, y| (x - 1, height - y));
+}
+
+fn border_radius(
+    img: &mut ImageBuffer<Rgba<u8>, Vec<u8>>,
+    r: u32,
+    coordinates: impl Fn(u32, u32) -> (u32, u32),
+) {
+    if r == 0 {
         return;
     }
+    let r0 = r;
+
+    // 16x antialiasing: 16x16 grid creates 256 possible shades, great for u8!
+    let r = 16 * r;
+
+    let mut x = 0;
+    let mut y = r - 1;
+    let mut p: i32 = 2 - r as i32;
+
+    // ...
```

Comment:
I had left this comment in accidentally, it doesn’t say anything; I think it used to be a marker where the main code should go.

Reply:
https://github.com/keiya01/og_image_writer/pull/26/commits/a963c4c827c15b46762f170681175cecdd13d323
I removed.